Syntax problems aren't caught by current testing

## Bug report

A bug came in recently (#878) that was a simple syntax issue. We should have caught this, but it slipped into a release. We have a patch already in the works, but I'm starting this issue to come up with an approach for preventing these types of problems from getting into releases in the future.

Maybe it's as simple as adding a linting step to our tests. If this were the case, I'd want it to be strict enough to catch syntax problems, but not so strict as to frustrate contributors with nit-picky suggestions.

Hello Jonathan,

I stumble upon this issue by accident, but since I've been investing linters for Python at work, I thought I might as well share my thoughts with you.
I think something like `pyflakes` would do the job just fine.

I retrieved the code to before the fix for #878 and I run `pyflakes` :
```
[maxime@nepture jrnl]$ time pyflakes jrnl/
jrnl/install.py:4: 'getpass' imported but unused
jrnl/install.py:10: '.Journal.PlainJournal' imported but unused
jrnl/install.py:11: '.EncryptedJournal.EncryptedJournal' imported but unused
jrnl/FolderJournal.py:5: '.Entry' imported but unused
jrnl/__init__.py:3: 'os' imported but unused
jrnl/EncryptedJournal.py:13: 'typing.Optional' imported but unused
jrnl/Journal.py:4: '.util' imported but unused
jrnl/time.py:60: use ==/!= to compare str, bytes, and int literals
jrnl/plugins/yaml_exporter.py:5: 'os' imported but unused
jrnl/plugins/jrnl_importer.py:5: '..util' imported but unused
jrnl/plugins/jrnl_importer.py:18: local variable 'old_entries' is assigned to but never used
jrnl/plugins/markdown_exporter.py:5: 'os' imported but unused

real	0m0,215s
user	0m0,192s
sys	0m0,022s
```
As you can see, it's crazy fast, and I would have caught the error in `time.py` (plus others things).
One thing to note thought, it only parses the source files, it doesn't note import them (that why it's so fast).

While investigating, I noticed that the Makefile already as a "lint" entry using `flake8`, but it's not used in the CI.
Since `flake8` use `pyflakes` amongst other libs, the following is similar to the command I run above :
```
[maxime@nepture jrnl]$ flake8 jrnl/ --select F 
jrnl/install.py:4:1: F401 'getpass' imported but unused
jrnl/install.py:10:1: F401 '.Journal.PlainJournal' imported but unused
jrnl/install.py:11:1: F401 '.EncryptedJournal.EncryptedJournal' imported but unused
jrnl/FolderJournal.py:5:1: F401 '.Entry' imported but unused
jrnl/__init__.py:3:1: F401 'os' imported but unused
jrnl/EncryptedJournal.py:13:1: F401 'typing.Optional' imported but unused
jrnl/Journal.py:4:1: F401 '.util' imported but unused
jrnl/time.py:60:8: F632 use ==/!= to compare str, bytes, and int literals
jrnl/plugins/yaml_exporter.py:5:1: F401 'os' imported but unused
jrnl/plugins/jrnl_importer.py:5:1: F401 '..util' imported but unused
jrnl/plugins/jrnl_importer.py:18:9: F841 local variable 'old_entries' is assigned to but never used
jrnl/plugins/markdown_exporter.py:5:1: F401 'os' imported but unused
```
The thing is, since Black is also used, there's no need for the part of flake8 that check the style of the code (which is `pycodestyle`). 

Here's what I propose : replacing flake8 by pyflakes, and running the lint task in the CI.
> in the Makefile
```
lint: ## pyflakes
	poetry run pyflakes .
```

> in the travis.yml
```
  include:
    - name: Lint, via Black
      python: 3.8
      script:
        - black --version
        - black --check . --verbose --diff
        - make lint

```

Another option is to use `Pylint`, with is a lot more powerful, but also a lot slower and requires a lot of tuning so I doesn't get in your way.

What's your thought on this?

--- a/.travis.yml
+++ b/.travis.yml
@@ -62,11 +62,15 @@ jobs:
     - python: nightly
 
   include:
-    - name: Lint, via Black
+    - name: Misc Linting
       python: 3.8
       script:
+        - poetry --version
+        - poetry check
         - black --version
-        - black --check . --verbose --diff
+        - black --check --diff .
+        - pyflakes --version
+        - pyflakes .
 
     # Python 3.6 Tests
     - name: Python 3.6 on Linux

--- a/Makefile
+++ b/Makefile
@@ -13,11 +13,12 @@ clean:
 html:
 	poetry run mkdocs serve
 
-format: ## check style with flake8
-	poetry run black features jrnl
+format: ## Format files to match style
+	poetry run black .
 
-lint: ## check style with flake8
-	poetry run flake8 jrnl features --ignore E501
+lint: ## Check style with various tools
+	poetry run black --check --diff .
+	poetry run pyflakes .
 
 test: ## Run behave tests
 	poetry run behave

--- a/features/steps/core.py
+++ b/features/steps/core.py
@@ -10,13 +10,10 @@ try:
 except ImportError:
     import parsedatetime as pdt
 import time
-from codecs import encode, decode
 import os
 import ast
-import json
 import yaml
 import keyring
-import tzlocal
 import shlex
 import sys
 from pathlib import Path
@@ -203,7 +200,6 @@ def check_output(context, text=None):
 @then('the output should contain "{text}" in the local time')
 def check_output_time_inline(context, text):
     out = context.stdout_capture.getvalue()
-    local_tz = tzlocal.get_localzone()
     date, flag = CALENDAR.parse(text)
     output_date = time.strftime("%Y-%m-%d %H:%M", date)
     assert output_date in out, output_date
@@ -230,7 +226,7 @@ def check_output_inline(context, text=None, text2=None):
 @then("the error output should contain")
 @then('the error output should contain "{text}"')
 @then('the error output should contain "{text}" or "{text2}"')
-def check_output_inline(context, text=None, text2=None):
+def check_error_output_inline(context, text=None, text2=None):
     text = text or context.text
     out = context.stderr_capture.getvalue()
     assert text in out or text2 in out, text or text2
@@ -299,7 +295,6 @@ def check_journal_entries(context, number, journal_name="default"):
 
 @when("the journal directory is listed")
 def list_journal_directory(context, journal="default"):
-    files = []
     with open(install.CONFIG_FILE_PATH) as config_file:
         config = yaml.load(config_file, Loader=yaml.FullLoader)
     journal_path = config["journals"][journal]

--- a/jrnl/EncryptedJournal.py
+++ b/jrnl/EncryptedJournal.py
@@ -10,7 +10,6 @@ import sys
 import os
 import base64
 import logging
-from typing import Optional
 
 
 log = logging.getLogger()

--- a/jrnl/FolderJournal.py
+++ b/jrnl/FolderJournal.py
@@ -2,7 +2,6 @@
 # encoding: utf-8
 
 from __future__ import absolute_import, unicode_literals
-from . import Entry
 from . import Journal
 import codecs
 import os

--- a/jrnl/__init__.py
+++ b/jrnl/__init__.py
@@ -1,7 +1,5 @@
 #!/usr/bin/env python
 
-import os
-
 try:
     from .__version__ import __version__
 except ImportError:

--- a/jrnl/cli.py
+++ b/jrnl/cli.py
@@ -365,7 +365,7 @@ def run(manual_args=None):
     try:
         journal = open_journal(journal_name, config)
     except KeyboardInterrupt:
-        print(f"[Interrupted while opening journal]", file=sys.stderr)
+        print("[Interrupted while opening journal]", file=sys.stderr)
         sys.exit(1)
 
     # Import mode

--- a/jrnl/install.py
+++ b/jrnl/install.py
@@ -1,14 +1,11 @@
 #!/usr/bin/env python
 
 import glob
-import getpass
 import os
 import xdg.BaseDirectory
 from . import util
 from . import upgrade
 from . import __version__
-from .Journal import PlainJournal
-from .EncryptedJournal import EncryptedJournal
 from .util import UserAbort, verify_config
 import yaml
 import logging

--- a/jrnl/plugins/jrnl_importer.py
+++ b/jrnl/plugins/jrnl_importer.py
@@ -2,7 +2,6 @@
 # encoding: utf-8
 
 import sys
-from .. import util
 
 
 class JRNLImporter:
@@ -15,7 +14,6 @@ class JRNLImporter:
         """Imports from an existing file if input is specified, and
         standard input otherwise."""
         old_cnt = len(journal.entries)
-        old_entries = journal.entries
         if input:
             with open(input, "r", encoding="utf-8") as f:
                 other_journal_txt = f.read()

--- a/jrnl/plugins/markdown_exporter.py
+++ b/jrnl/plugins/markdown_exporter.py
@@ -2,7 +2,6 @@
 # encoding: utf-8
 
 from .text_exporter import TextExporter
-import os
 import re
 import sys
 from ..util import WARNING_COLOR, RESET_COLOR

--- a/jrnl/plugins/template.py
+++ b/jrnl/plugins/template.py
@@ -6,9 +6,7 @@ EXPRESSION_RE = r"[\[\]():.a-zA-Z0-9_]*"
 PRINT_RE = r"{{ *(.+?) *}}"
 START_BLOCK_RE = r"{% *(if|for) +(.+?) *%}"
 END_BLOCK_RE = r"{% *end(for|if) *%}"
-FOR_RE = r"{{% *for +({varname}) +in +([^%]+) *%}}".format(
-    varname=VAR_RE, expression=EXPRESSION_RE
-)
+FOR_RE = r"{{% *for +({varname}) +in +([^%]+) *%}}".format(varname=VAR_RE)
 IF_RE = r"{% *if +(.+?) *%}"
 BLOCK_RE = r"{% *block +(.+?) *%}((?:.|\n)+?){% *endblock *%}"
 INCLUDE_RE = r"{% *include +(.+?) *%}"

--- a/jrnl/plugins/yaml_exporter.py
+++ b/jrnl/plugins/yaml_exporter.py
@@ -2,7 +2,6 @@
 # encoding: utf-8
 
 from .text_exporter import TextExporter
-import os
 import re
 import sys
 from ..util import WARNING_COLOR, ERROR_COLOR, RESET_COLOR
@@ -19,10 +18,9 @@ class YAMLExporter(TextExporter):
         """Returns a markdown representation of a single entry, with YAML front matter."""
         if to_multifile is False:
             print(
-                "{}ERROR{}: YAML export must be to individual files. Please \
-                specify a directory to export to.".format(
-                    ERROR_COLOR, RESET_COLOR, file=sys.stderr
-                )
+                f"{ERROR_COLOR}ERROR{RESET_COLOR}: YAML export must be to individual files. Please \
+                specify a directory to export to.",
+                file=sys.stderr,
             )
             return
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -33,9 +33,9 @@ ansiwrap = "^0.8.4"
 [tool.poetry.dev-dependencies]
 behave = "^1.2"
 mkdocs = "^1.0"
-flake8 = "^3.7"
 black = {version = "^19.10b0",allow-prereleases = true}
 toml = "^0.10.0"
+pyflakes = "^2.2.0"
 
 [tool.poetry.scripts]
 jrnl = 'jrnl.cli:run'

